### PR TITLE
Batch query services

### PR DESCRIPTION
**Pull request by akaplya**
## Problem
Batch query services. This document specifies how input/output arguments should look like to implement batch query services

### CODE REVIEW

File: design-documents/batch-query-services.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
+#Batch query services
+
+## Purpose
+
+The main reason for such change is performance.
+The system triggers multiple queries during an operation (page load, service call). 
+Instead of making several requests for data retrieval,
+it would be more efficient to assemble all requests and send them as the single call to query service.
```
Comment (kokoc) [inline design-documents/batch-query-services.md]: I think it's worth to say that we are assembling multiple requests of the same type

File: design-documents/batch-query-services.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
+#Batch query services
```
Comment (kokoc) [inline design-documents/batch-query-services.md]: I think we need to add error reporting section as it's very important part for batch APIs.

File: design-documents/batch-query-services.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+## Operations output
+
+Because we may receive multiple requests to one service it makes sense to return the results segregated per the request.
+This decision allows us to reduce effort on the client to recognise part of the result that belongs to it.
+As a consequence, the client should not keep state (criteria) between requests. 
+
+```php
+public class ResultContainer
+{
+    public function getCriteria() : CrtiteriaInterface
```
Comment (kokoc) [inline design-documents/batch-query-services.md]: I think that criteria is useless in api response. Clients may rely on query  position which is much simpler comparing to searchCriteria comparison

File: design-documents/batch-query-services.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+Even without internal code optimization such approach saves time that required for the service components bootstrap.
+* Such an approach saves time that required for the service components bootstrap.
+* The service implementation can be changed to guarantee better performance for batch operations.
+
+## Arguments of operation
+
+A service that manipulates with a single request could have a straightforward interface:
+```php
+    /**
+     * @param int $productId
```
Comment (kandy) [inline design-documents/batch-query-services.md]: should be int[] type

File: design-documents/batch-query-services.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+     * @return array
+     */
+    public function getProductPrice(array $args) : array
+```
+
+We can use object representation for input arguments to achieve better transparency and enforce strict types check.
+```php
+class ProductPriceSearchCriteria implements CriteriaInterface
+{
+    public function getProductId() : int;
```
Comment (kandy) [inline design-documents/batch-query-services.md]: should be int[] type

File: design-documents/batch-query-services.md (hunk 2 above)
Comment (akaplya) [reply design-documents/batch-query-services.md]: done

File: design-documents/batch-query-services.md (hunk 4 above)
Comment (akaplya) [reply design-documents/batch-query-services.md]: Not exactly correct. actually, this proposal about how to satisfy multiple clients of a single service. 
Arguments segmentation really matters in this case.

File: design-documents/batch-query-services.md (hunk 5 above)
Comment (akaplya) [reply design-documents/batch-query-services.md]: see the previous comment.

File: design-documents/batch-query-services.md (hunk 1 above)
Comment (akaplya) [reply design-documents/batch-query-services.md]: fixed

File: design-documents/batch-query-services.md (hunk 3 above)
Comment (akaplya) [reply design-documents/batch-query-services.md]: added options to discuss

File: design-documents/batch-query-services.md (hunk 4 above)
Comment (kandy) [reply design-documents/batch-query-services.md]: So, do you think that this service should not use batch? Looks like our Technical Vision say that all services should be batched

**Comment by AlexMaxHorkun** [comment]
@akaplya Perhaps we don't have burden our (and 3rd party services) with these complex method definition requirements? Please check out https://github.com/magento/magento2ce/pull/4366 we could use deferred result proxies to keep the same method definitions but to postpone/make asynchronous operations

File: design-documents/batch-query-services.md [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    public function getCriteria() : CrtiteriaInterface
+    public function getResult() : array
+}
+```
+
+Currently, the interface looks more clear.
+```php
+    /**
+     * @param ProductPriceSearchCriteria[] $args
+     * @return CrtiteriaInterface[]
```
Comment (buskamuza) [inline design-documents/batch-query-services.md]: Maybe `ResultContainer`?

File: design-documents/batch-query-services.md [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+     */
+    public function getProductPrice(array $args) : array
+```
+For some cases, requested information that returned with the response may help to build a stateless scenario.
+For instance, a widget which shows weather for the detected area and refreshes forecast every 5 minutes.
+
+Cons service always has to return request arguments increasing the size of the message that we are transfering through the network
```
Comment (buskamuza) [inline design-documents/batch-query-services.md]: "Cons:"

File: design-documents/batch-query-services.md [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+]
+```
+
+### 2. Return execution result in the order of received requests
+
+Results can be mapped to requests according to the order of incoming arguments.
+With this approach, we do not have to return request criteria with results.
+Client assumes that the order of received result containers is equal to the order of sent arguments.
+Service implementation has to guarantee this sort order.
+This approach does not work with services that returns
```
Comment (buskamuza) [inline design-documents/batch-query-services.md]: that returns what?

File: design-documents/batch-query-services.md [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+         "minimalPrice":"8.56"
+      }
+   }
+]
+```
+
+## Error reporting
+
+An error may occur on overall service execution or during processing one of the received criteria.
+This is up to a developer how the service tolerates and handles runtime exceptions during.
```
Comment (buskamuza) [inline design-documents/batch-query-services.md]: during...

**Pull request merged**